### PR TITLE
fix(store): capacity limit for persistent message store

### PR DESCRIPTION
**Pull request by kaiserd**
This PR provides a simple solution to https://github.com/status-im/infra-status/issues/5 and https://github.com/status-im/nim-waku/issues/767

Currently, the DB grows boundless, and, while only a subset of messages is loaded into the in-memory store, an `ORDER-BY` operation is run on the whole DB.
This might lead to very long loading times https://github.com/status-im/infra-status/issues/5 

This PR adds `numMessages` (a message counter) and `maxMessages` (max number of messages to be held in the sqlite DB) to the `WakuMessageStore`.
When `put`ting a message into the store, it checks if `maxMessages` is reached and if so, deletes a configurable percentage of oldest messages.

This PR introduces two new config options:

* `--db-capacity` the maximum number of messages stored in the DB
* `--db-truncate-percentage` the percentage of messages to be deleted once the max is reached

These are added for flexibility when testing on the test fleet and might be removed in the future.

This PR further changes the first part of the sqlite primary key from `senderTimestamp` to `receiverTimestamp`.
Rationale: Most costly sort operations are on `receiverTimestamp`, so this change increases performance.
I would suggest, building a dedicated index on `receiverTimestamp` as even with this change, sqlite has to build a temporary B-tree when ordering by `receiverTimestamp`.
(For this, I could introduce another issue. As we plan to update to a more efficient DB structure anyway, we might wait for the respective issue.)

## Alternatives

These are alternatives to solving the issues mentioned above:

* don't change the implementation but simply run a chron job truncating the DB from time to time
* trigger the `trunctate` proc introduced by this PR within a timer instead of on `put`.

We are planning to update the DB structure to make it more efficient, which will solve these issues, too.
So this PR is more of a quick fix.

cc @jakubgs 

## Todo

These todo's should be completed before making the PR mergeable / "ready for review".

* [x] fix test error : reverted the primary key change for now
* [x] add tests
* ~adapt the migration test to the new primary key~ no primary key update in this PR
* [ ] update documentation

### CODE REVIEW

**Review by jakubgs — COMMENTED** [review]
Wouldn't setting `storeCapacity` to a value lower than `dbCapacity` invalidate the first setting?
If so, should that at least trigger some warning if not an error?

Seems like both of those parameters are for the same thing, but work differently, that could be quite confusing. Also, doesn't this approach mean that at times we'd have less messages than `dbCapacity`accounts for?

**Comment by kaiserd** [comment]
Thank you for the feedback.

> Wouldn't setting `storeCapacity` to a value lower than `dbCapacity` invalidate the first setting?
> If so, should that at least trigger some warning if not an error?

I could add such an error. As an alternative, the two new options could also be removed and automatically set to `storeCapacity * 2` and `20` for example.

> Seems like both of those parameters are for the same thing, but work differently, that could be quite confusing.

One is for in-memory, the other for persistence. I agree, this is confusing with the current naming.  Most likely, it would be better to just have one option controlling both. With these paramaters, I wanted to give more flexibility in a testing phase.

>  Also, doesn't this approach mean that at times we'd have less messages than `dbCapacity` accounts for?

Yes. I think, deleting in batches makes more sense because of efficiency.
But yes, this adds to the confusion you pointed out.

My suggestion would be not exposing `dbCapacty` and `dbTruncatePercentage`. This would solve the confusion issue you pointed out. Maybe it would make sense, to do a few tests checking which values we should choose relateive to `storeCapacity`. Also, we could add a sentence in the doc explaining that `storeCapacity` just refers to in-memory storage.
(Will discuss this in today's meeting.)

**Comment by kaiserd** [comment]
I revertet the primary key for now, because it made problems with the current migration test.

**Comment by kaiserd** [comment]
> If we want query cache then it should be done using something dedicated, like memcached, or similar.

Regarding this, I was thinking about using something like Redis because most of the operations happen on the in-memory store and the DB is just for persisting, which Redis can do, too, on regular intervals.
Also, afaik, the focus for the store is more on efficiency than transactional safety and it should not be a big deal if a few messages get lost on a few nodes, because others would still likely hold the respective messages. I will soon have a 1:1 with @jm-clius planning our store efficiency issue.

**Comment by jakubgs** [comment]
>This PR is just a quick-fix for the very long loading times.

Of course, just trying to clear some things up.

>Regarding this, I was thinking about using something like Redis because most of the operations happen on the in-memory store and the DB is just for persisting, which Redis can do, too, on regular intervals.

Yeah, agree, Redis is a nice choice, I quite like it as a cache.

**Comment by jakubgs** [comment]
@oskarth @s1fr0 @staheri14 @jm-clius can we please get some reviews going?

This issue is causing nodes to go down at every deployment and config change for an hour or more now.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (jm-clius) [inline waku/v2/node/storage/message/waku_message_store.nim:75]: A bit of a nitpick, but doesn't `truncate` in SQL terms usually mean deleting _all_ rows from a table (not just the oldest)? In other words, if we want to add such a function in future we may run into clashing terminology. Perhaps calling this something closer to `deleteOldest` or `resize` could avoid confusion in future.

**Review by jm-clius — APPROVED** [review]
Thanks, @kaiserd. I think this is a mergeable minimal fix for the DB capacity issue! As you mentioned in the conversation above, we can work on improving our DB solution and efficiency in subsequent PRs.

**Comment by oskarth** [comment]
Please mark PR as "ready for review" once it is ready for proper review

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (jm-clius) [reply waku/v2/node/storage/message/waku_message_store.nim:75]: Ah, yes! That's on me. :)

**Comment by kaiserd** [comment]
I updated the PR addressing @jakubgs' feedback.
Further, I removed the `limit` parameter from the `getAll` method.
The only configurable limit is now `storeCapacity` which can be set via `--store-capacity`.
Also, auto-delete now happens on node start, too, when loading a DB that exceeds the limit already.

Auto-delete now works like this (I added this explanation as a comment to the `WakuMessageStore` type):

The sqlite DB will store up to `storeCapacity` * `maxStoreOverflow` messages. `maxStoreOverflow` is a `const` set to `1.2`.
If more messages are stored, the messages are sorted by `receiverTimestamp` and the oldest ones are deleted,
bringing the total number of stored messages back to `storeCapacity*maxStoreOverflow/2`. The rationale for batch deleting is efficiency.
We keep half of the overflow window because we delete the oldest messages with respect to `receiverTimestamp` instead of `senderTimestamp`. `ReceiverTimestamp` is guaranteed to be set, while senders could omit setting `senderTimestamp`. However, `receiverTimestamp` can differ from node to node for the same message.
So, sorting by `receiverTimestamp` might (slightly) prioritize some actually older messages and we compensate that by keeping half of the overflow window.

(I got delayed because I ran into a nasty bug. I forgot to adapt the signature of `getAll` in `message_store` which was hard to find due to Nim's dynamic dispatch.)

I also commented out a test for the old limit parameter.

This PR will become mergeable / "ready for review" after adding/adapting tests and adding docu which I will do asap.

I will open a new issue for
* introducing an index on `receiverTimestamp`, and
* using [sqlite3 vaccum](https://www.sqlite.org/lang_vacuum.html)

for increasing efficiency.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (staheri14) [inline waku/v2/node/storage/message/waku_message_store.nim:42]: It seems an expensive query that can take a lot of time to conclude (of course depends on the internals of the db, but in the worst case it can take `n`*`m` comparisons where `n` is the DB size and `m` is the selected portion), how about selecting the portion you want, then erasing the entire table and replacing its content with that already selected portion.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (staheri14) [inline waku/v2/node/storage/message/waku_message_store.nim:58]:
```suggestion
  ##  - Payload stored as a blob
```

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (staheri14) [inline waku/v2/node/storage/message/waku_message_store.nim:29]: > bringing the total number of stored messages back to `storeCapacity*maxStoreOverflow/2`. The rationale for batch deleting is efficiency. 

I assume the `storeCapacity` indicates the in-memory limit, right? if yes, then I think the `maxStoreOverflow` should be a value larger than `2`, otherwise, while truncating, you may end up deleting DB messages that are still available in the memory. 

For example, for `maxStoreOverflow = 1.3`, the persisted db messages will be `storeCapacity`*0.65, which means the db does not contain all the in-mem messages.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (staheri14) [inline waku/v2/node/storage/message/waku_message_store.nim:30]:
```suggestion
  # We keep half of the db capacity because we delete the oldest messages with respect to `receiverTimestamp` instead of `senderTimestamp`.
```

**Review by staheri14 — COMMENTED** [review]
Looks fine in general. Please see my comments inline. 
A suggestion about the `storeCapacity` name, you may want to consider `inMemCapacity` instead (might be a bit less confusing and more self-explanatory).

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (jm-clius) [reply waku/v2/node/storage/message/waku_message_store.nim:42]: We are indeed planning several DB and efficiency improvements for subsequent PRs.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (jm-clius) [reply waku/v2/node/storage/message/waku_message_store.nim:29]: Right, if the idea is to only remove half of the _overflow_ (not half of the entire capacity), shouldn't the formula for total messages after deletion be something like `storeCapacity*(1 + (maxStoreOverflow - 1)/2)`. A bit of a complicated way to express this though. Perhaps it's easier to just determine the absolute number of overflow messages allowed `(maxStoreOverflow*storeCapacity - storeCapacity)` when initiating the store and always removing half of that?

**Comment by kaiserd** [comment]
Thank you very much for your feedback @staheri14 . I will add tests and then address your feedback.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (kaiserd) [reply waku/v2/node/storage/message/waku_message_store.nim:29]: @staheri14 Yes. Good point. The store does not sync between DB and in-memory store.
This is also one of the points we plan to address in the future.
The currently implemented formula is as @jm-clius stated.
Only half of the overflow window gets deleted. The full `storeCapacity` and half of the overflow window are kept.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (kaiserd) [reply waku/v2/node/storage/message/waku_message_store.nim:42]: Yes. In the current implementation, most of the messages are kept and only a comparatively small part is deleted.
So, selecting the messages that are kept would take most of the time.
An exception would be if a very large DB of an older version is loaded.

My suggestion for making this query more efficient would be adding an index on `receiverTimestamp`,
which would reduce the cost for the select part to log(m).
I will open an issue for this (+ vacuum) as soon as this issue is closed.

While Index + vacuum will be addressed timely, future steps we envisioned are:

* pure SQL store
  - large storage capacity but slow
  - in-line with status-desktop requirement of keeping messages for 30 day
* adding an efficient cache

I could also add a store efficiency roadmap issue tracking these future goals.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (kaiserd) [reply waku/v2/node/storage/message/waku_message_store.nim]: We actually keep capactiy + half of the overflow window. See comment above.
I will write this more clearly.

**Comment by kaiserd** [comment]
I pushed a new commit addressing feedback and adding tests.

There is still an issue I have to figure out before opening the PR.
In the [test]() I added, when setting capacity and overload to a larger number, e.g.

```
let
  capacity = 1000
  overload = 1000
```

the messages do not get deleted properly anymore.
I added a debug statement printing the number of messages actually in the DB after the deletion.
This could be removed later for efficiency, because the count has to go over the DB one more time.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (kaiserd) [reply waku/v2/node/storage/message/waku_message_store.nim]: > Perhaps it's easier to just determine the absolute number of overflow messages allowed (maxStoreOverflow*storeCapacity - storeCapacity) when initiating the store and always removing half of that?

I now added a `deleteWindow` to `WakuMessageStore`.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (staheri14) [inline waku/v2/node/storage/message/waku_message_store.nim:20]: I suggest using capital case instead (it is easier to distinguish local vars from const)
```suggestion
const MAX_STORE_OVERFLOW = 1.3
```

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (staheri14) [inline waku/v2/node/storage/message/waku_message_store.nim:27]: Thanks, this new formulation addresses my past comment

**Review by staheri14 — APPROVED** [review]
LGTM! approved it though if it is still a WIP and needed another review, ping me again.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (kaiserd) [reply waku/v2/node/storage/message/waku_message_store.nim:20]: I did not use capital case because of [NEP1](https://nim-lang.org/docs/nep1.html#introduction-naming-conventions), and I did not find anything overriding this in [status style guide](https://status-im.github.io/nim-style-guide/00_introduction.html).
No strong opinion on my side. Wdyt?

**Comment by kaiserd** [comment]
Thank you for reviewing the PR again :). I will open the PR as soon as I figure out the problem in the test described above.

**Comment by kaiserd** [comment]
The PR is now open :).

The issue in the test mentioned above occurred because the content of the messages in the test was only a single byte, which wrapped around when storing more than 256 messages.
Since the id only depends on the message digest, which in turn depends on the message payload and `pubsubTopic` (only one in the test), sets of messages shared the same id.

A second issue was the second-resolution of the receiverTimestamp which I overlooked.
A comment in the test said that 1ms of waiting would result in different timestamps. I removed that comment and the 1ms wait.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (jm-clius) [reply waku/v2/node/storage/message/waku_message_store.nim:20]: Yeah, no strong opinions either, other than that we've certainly not been consistent so far. :/ I suggest we go with the linked Nim convention and specifically PascalCase - this is also what `nim-libp2p` uses. A good exercise for our next focused refactoring effort. Thanks for highlighting, @staheri14.

**Review by jm-clius — APPROVED** [review]
LGTM! Thanks :)

**Comment by jm-clius** [comment]
@kaiserd can this be merged, or are you still waiting for one more review?

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (staheri14) [inline waku/v2/node/storage/message/waku_message_store.nim:20]: Although it would never happen probably, it is prudent to add a comment that this value must be larger than `1`, otherwise the math at line 98, i.e., 
> let deleteWindow = int(float(storeMaxLoad - storeCapacity) / 2)
will yield a negative value.

**Review by staheri14 — APPROVED** [review]
LGTM! feel free to apply my last comment in your subsequent PRs, nothing urgent.

File: waku/v2/node/storage/message/waku_message_store.nim
Comment (kaiserd) [reply waku/v2/node/storage/message/waku_message_store.nim:20]: I will do this. Thanks :).

**Pull request merged**
